x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,154 @@
+	start           time.Time
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	acker := &EventACKTracker{
+		ID:         ackerIDCounter.Inc(),
+		ctx:        ctx,
+		cancel:     cancel,
+		deletionWg: deletionWg,
+	}
+
+	go func() {
+		t := time.NewTicker(500 * time.Microsecond)
+		defer t.Stop()
+
+		for {
+			<-t.C
+			if !acker.FullyTracked() {
+				continue
+			}
+
+			acker.cancelAndFlush()
+			return
+		}
+	}()
+
+	return acker
+}
+
+func (a *EventACKTracker) cancelAndFlush() {
+	a.deleteSQS()
+	a.cancel()
+}
+
+// MarkSQSProcessedWithData Every call after the first one is a no-op
+func (a *EventACKTracker) MarkSQSProcessedWithData(msg *types.Message, publishedEvent uint64, receiveCount int, start time.Time, processingErr error, handles []s3ObjectHandler, keepaliveCancel context.CancelFunc, keepaliveWg *sync.WaitGroup, msgHandler sqsProcessor, log *logp.Logger) {
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if a.eventsToBeAcked.Load() > 0 {
+		return
+	}
+
+	a.msg = msg
+	a.eventsToBeAcked.Store(publishedEvent)
+	a.ReceiveCount = receiveCount
+	a.start = start
+	a.processingErr = processingErr
+	a.Handles = handles
+	a.keepaliveCancel = keepaliveCancel
+	a.keepaliveWg = keepaliveWg
+	a.msgHandler = msgHandler
+	a.log = log
+}
+
+func (a *EventACKTracker) FullyTracked() bool {
+	eventsToBeTracked := a.eventsToBeAcked.Load()
+	if eventsToBeTracked == 0 {
+		return false
+	}
+
+	// This is eating its own tail: we should check for dropped+published, but then we won't wait for acked.
+	// Acked might not be equal to published?
+	return a.eventsAcked.Load() == eventsToBeTracked
+}
+
+// deleteSQS delete related SQS message
+func (a *EventACKTracker) deleteSQS() {
+	// Stop keepalive visibility routine before deleting.
+	a.keepaliveCancel()
+	a.keepaliveWg.Wait()
+
+	err := a.msgHandler.DeleteSQS(a.msg, a.ReceiveCount, a.processingErr, a.Handles)
+	if err != nil {
+		a.log.Warnw("Failed deleting SQS message.",
+			"error", err,
+			"message_id", *a.msg.MessageId,
+			"elapsed_time_ns", time.Since(a.start))
+	} else {
+		a.log.Debugw("Success deleting SQS message.",
+			"message_id", *a.msg.MessageId,
+			"elapsed_time_ns", time.Since(a.start))
+	}
+
+	a.deletionWg.Done()
+
+}
+
+// ACK increments the number of eventsAcked.
+func (a *EventACKTracker) ACK() {
+	a.eventsAcked.Inc()
+}
+
+// NewEventACKHandler returns a beat ACKer that can receive callbacks when
+// an event has been ACKed an output. If the event contains a private metadata
+// pointing to an eventACKTracker then it will invoke the trackers ACK() method
+// to decrement the number of pending ACKs.
+func NewEventACKHandler() beat.EventListener {
```

Comment:
Can `NewEventACKHandler` function stay inside libbeat?

Reply:
we already have `awscommon.EventACKTracker` in `github.com/elastic/beats/v7/x-pack/libbeat/common/aws`
This akcer is specific for sqs-s3 notifcations